lib/services/local/generic/resource_adapter.go

```diff
@@ -0,0 +1,49 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package generic
+
+import (
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+	"github.com/gravitational/teleport/api/types"
+)
+
+// resourceMetadataAdapter is an adapter for RFD 153-style resources. For inner types T which implement
+// types.ResourceMetadata, the type resourceMetadataAdapter will implement both types.ResourceMetadata and Resource.
+//
+// resourceMetadataAdapter is similar to resource153ToLegacyAdapter which can be found api/types/resource_153.go,
+// yet resource153ToLegacyAdapter is different in several ways:
+// - it isn't generic,
+// - it is unexported,
+// - it implements the entire types.Resource interface,
+// - it makes certain opinionated implementation choices, including calling panic().
```

Comment:
```suggestion
// - it makes certain opinionated implementation choices, including panicking if
//   the underlying resource is nil
```

If the suggestion is not correct, then when does it panic?

Reply:
Well, that kind of stands out:

```go
func (r *resource153ToLegacyAdapter) SetSubKind(subKind string) {
	panic("interface Resource153 does not implement SetSubKind")
}
```

https://github.com/gravitational/teleport/blame/cfabeb41d3800cf4b3431ddd5fdc3bae81158102/api/types/resource_153.go#L207-L209

Reply:
Is that what you meant by panicking, then?

The sentence right before says: "yet resource153ToLegacyAdapter is different in several ways". If it's the same panic as resource153ToLegacyAdapter, and just that, then maybe cut this last item?

Reply:
I'm bit confused. `resourceMetadataAdapter` doesn't panic, `resource153ToLegacyAdapter` does. This is a list of the ways the two types are different, and it is one of the differences.

Reply:
I was trying to say I don't get, based on the comment, when this can panic. Up to you if you want to tweak it.